thunder/executors/nvfuserex_impl.py

```diff
@@ -2732,6 +2737,268 @@ def embedding(
+    # gather/ take_along_axis on the input tensor
+    ne = fd.ops.ne(nv_target, nv_ignore_index)
+    where_0 = fd.ops.where(ne, nv_target, zero_scalar)
+    where = fd.ops.broadcast_in_dim(where_0, shape=[nv_target.shape()[-1], 1], broadcast_dims=[0])
+    gather0 = fd.ops.take_along_axis(nv_a, where, dim=1)
+    gather = fd.ops.reshape(gather0, new_shape=[nv_target.shape()[-1]])
+
+    # compute log (sum(exp(A - max(A, dim=1), dim=1))
+    max = fd.ops.max(nv_a, 1)
+    max2 = fd.ops.broadcast_in_dim(max, shape=[nv_target.shape()[-1], 1], broadcast_dims=[0])
+    sub = fd.ops.sub(nv_a, max2)
+    exp = fd.ops.exp(sub)
+    sum_1 = fd.ops.sum(exp, 1)
+    log = fd.ops.log(sum_1)
+
+    # compute S = POST_GATHER(A) - max(A, dim=1) - log(sum(exp(A - max(A, dim=1), dim=1))
+    gather_sub = fd.ops.sub(gather, max)
+    log_softmax_post_gather = fd.ops.sub(gather_sub, log)
+
+    # set the values for ignore index to 0
+    neg = fd.ops.neg(log_softmax_post_gather)
+    where_1 = fd.ops.where(ne, neg, zero_scalar)
+
+    # sum_2_cvt computes the number of valid indices
+    sum_2 = fd.ops.sum(ne)
+    sum_2_cvt = fd.ops.cast(sum_2, dtype=DataType.Float)
+
+    # compute the mean
+    sum_3 = fd.ops.sum(where_1)
+    div = fd.ops.div(sum_3, sum_2_cvt)
+    return div, max, log, sum_2_cvt
+
+
+nv_cross_entropy_fwd = ex.register_operator(
+    "nv_cross_entropy_fwd",
+    meta=cross_entropy_fwd_meta,
+)
+register_supported(nv_cross_entropy_fwd.id, cross_entropy_fwd, None)
+
+
+def cross_entropy_bwd_meta(
+    g: TensorLike,
+    a: TensorLike,
+    *,
+    target: TensorLike,
+    a_max: TensorLike,
+    max_log_sum_exp: TensorLike,
+    valid_indices: TensorLike,
+    ignore_index: int = -100,
+    label_smoothing: float = 0.0,
+) -> Any:
+    return TensorProxy(like=a)
+
+
+def cross_entropy_bwd(
+    g: TensorLike,
+    a: TensorLike,
+    *,
+    target: TensorLike,
+    a_max: TensorLike,
+    max_log_sum_exp: TensorLike,
+    valid_indices: TensorLike,
+    ignore_index: int = -100,
+    label_smoothing: float = 0.0,
+    fd: FusionDefinition,
+    lc_to_nv_map: dict,
+) -> Any:
+    nv_a = getnv(a, fd, lc_to_nv_map)
+    nv_target = getnv(target, fd, lc_to_nv_map)
+    nv_ignore_index = getnv(ignore_index, fd, lc_to_nv_map)
+    nv_g = getnv(g, fd, lc_to_nv_map)
+    nv_valid_indices = getnv(valid_indices, fd, lc_to_nv_map)
+    nv_a_max = getnv(a_max, fd, lc_to_nv_map)
+    nv_max_log_sum_exp = getnv(max_log_sum_exp, fd, lc_to_nv_map)
+
+    zero = fd.define_scalar(0, dtype=DataType.Int)
+    one = fd.define_scalar(1, dtype=DataType.Int)
+
+    # scatter the gradients (negative) - this is backward of nll los
+    iotas = fd.ops.iota(nv_a.shape()[-1], zero, one, dtype=DataType.Int)
+    iotas_bcast = fd.ops.broadcast_in_dim(iotas, shape=nv_a.shape(), broadcast_dims=[nv_a.ndim - 1])
+    neg_gradients = fd.ops.neg(nv_g)
+    neg_gradients_mod = fd.ops.div(neg_gradients, nv_valid_indices)
```

Comment:
Again, could nv_valid_indices be zero?

Reply:
When there all entries in labels/targets is ignore index, this will still match PyTorch (tested out both fwd and bwd).
This is because `scattered_vals` become 0 and so does `softmax_mul_grad_sum`.